src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_cookiecontainer);
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            TaskCompletionSource<HttpResponseMessage> tcs = new TaskCompletionSource<HttpResponseMessage>();
+
+            // Create RequestState object and save current values of handler settings.
+            RequestState state = new RequestState();
+            state.Tcs = tcs;
+            state.CancellationToken = cancellationToken;
+            state.RequestMessage = request;
+            state.Handler = this;
+            state.CheckCertificateRevocationList = _checkCertificateRevocationList;
+            state.ServerCertificateValidationCallback = _serverCertificateValidationCallback;
+            state.WindowsProxyUsePolicy = this.WindowsProxyUsePolicy;
+            state.Proxy = this.Proxy;
+            state.ServerCredentials = _serverCredentials;
+
+            try
+            {
+                Task.Factory.StartNew(StartRequest, state);
+            }
+            catch (Exception ex)
+            {
+                throw new HttpRequestException("SendAsync", ex);
+            }
+
+            return tcs.Task;
+        }
+
+        private static uint ChooseAuthScheme(uint supportedSchemes)
+        {
+            foreach (uint authScheme in s_authSchemePriorityOrder)
+            {
+                if ((supportedSchemes & authScheme) != 0)
+                {
+                    return authScheme;
+                }
+            }
+
+            return 0;
+        }
+
+        private static void WinHttpStatusCallback(
+            IntPtr handle,
+            IntPtr context,
+            uint internetStatus,
+            IntPtr statusInformation,
+            uint statusInformationLength)
+        {
+            RequestState state;
+
+            // TODO: Use logging.
+
+            if (internetStatus != Interop.WinHttp.WINHTTP_CALLBACK_STATUS_REDIRECT &&
+                internetStatus != Interop.WinHttp.WINHTTP_CALLBACK_STATUS_SENDING_REQUEST)
+            {
+                return;
+            }
+
+            // Get request state object.
+            try
+            {
+                GCHandle gch = GCHandle.FromIntPtr(context);
+                state = (RequestState)gch.Target;
+            }
+            catch (Exception)
+            {
+                // TODO: We should log this but an exception here is pretty rare and
+                // is probably the result of a programming error.
+
+                // Closing the WinHTTP request handle will cancel the operation.
+                Interop.WinHttp.WinHttpCloseHandle(handle);
+
+                return;
+            }
+
+            if (internetStatus == Interop.WinHttp.WINHTTP_CALLBACK_STATUS_REDIRECT)
+            {
+                try
+                {
+                    var redirectUriString = Marshal.PtrToStringUni(statusInformation);
```

Comment:
The guideline confusingly paints `new` as good, and `OpenStandardInput()` as bad.  In https://github.com/stephentoub/corefx/commit/6daf1a66c693122f9bc07507255fa725d07dfd4a#commitcomment-11636599 we sort of evolved that a method name can be misleading, even with a type in it (could be a much more specific derived type, and the average reader wouldn't know that).

My understanding is:
* `new` is good
* hard-cast is probably good (`var foo = (Foo)obj;`)
* soft-cast might be good (`var foo = obj as Foo;`)
* Hiding a really long, but obvious, type name in a foreach is probably a violation that no one is going to complain about (`foreach (var kvp in someDictionary) { ... }`)

We should probably firm up the wording on that point to reduce the amount of grey area. (cc: @stephentoub, @ellism, @terrajobst)